lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -0,0 +1,243 @@
+            // generated.
+            auto xmrDefOp = getInnerRefTo(send.getBase().getDefiningOp());
+            // Record the remote reference op, that this ref value refers to.
+            reachingRefSendAt[send.getResult()] =
+                ArrayAttr::get(send.getContext(), {xmrDefOp});
+            opsToRemove.push_back(send);
+            return success();
+          })
+          .Case<InstanceOp>([&](auto inst) { return handleInstanceOp(inst); })
+          .Case<FConnectLike>([&](FConnectLike connect) {
+            // Ignore BaseType.
+            if (!connect.getSrc().getType().isa<RefType>())
+              return success();
+            // Get the dataflow value into the src.
+            if (auto remoteOpPath = getRemoteRefSend(connect.getSrc(), connect))
+              reachingRefSendAt[connect.getDest()] = remoteOpPath;
+            else
+              return failure();
+            opsToRemove.push_back(connect);
+            return success();
+          })
+          .Case<RefResolveOp>(
+              [&](auto resolve) { return handleRefResolve(resolve); })
+          .Default([&](auto) { return success(); });
+    };
+
+    DenseSet<Operation *> visitedModules;
+    // Traverse the modules in post order.
+    for (auto node : llvm::post_order(&instanceGraph)) {
+      auto module = cast<FModuleOp>(*node->getModule());
+      if (!visitedModules.insert(module).second)
+        continue;
+
+      LLVM_DEBUG(llvm::dbgs()
+                 << "\n Traversing module:" << module.moduleNameAttr());
+
+      for (Operation &op : module.getBodyBlock()->getOperations())
+        if (transferFunc(op).failed())
+          return signalPassFailure();
+
+      // Record all the RefType ports to be removed later.
+      for (size_t portNum = 0, e = module.getNumPorts(); portNum < e; ++portNum)
+        if (module.getPortType(portNum).isa<RefType>())
+          refPortsToRemoveMap[module].push_back(portNum);
+    }
+
+    // Now erase all the Ops and ports of RefType.
+    // This needs to be done as the last step to ensure uses are erased before
+    // the def is erased.
+    for (auto op : llvm::reverse(opsToRemove))
+      op->erase();
+    for (auto iter : refPortsToRemoveMap)
+      if (auto mod = dyn_cast<FModuleOp>(iter.getFirst()))
+        mod.erasePorts(iter.getSecond());
+      else if (auto inst = dyn_cast<InstanceOp>(iter.getFirst())) {
+        ImplicitLocOpBuilder b(inst.getLoc(), inst);
+        inst.erasePorts(b, iter.getSecond());
+        inst.erase();
+      }
+  }
+
+  ArrayAttr getRemoteRefSend(Value val, Operation *op) {
+    auto iter = reachingRefSendAt.find(val);
+    if (iter != reachingRefSendAt.end())
+      return iter->getSecond();
+    // The referenced module must have already been analyzed, error out if the
+    // dataflow at the child module is not resolved.
+    op->emitOpError(
+        "reference dataflow cannot be traced back to the remote read op");
+    return {};
+  }
```

Comment:
It would be nice if there was a way to avoid having to pass the `Operation *` as this is only really used in the failing path. It's also annoying to have to do the special casing here if the `val` is a block arg or a has a defining op. I'd consider doing it the latter way.

Reply:
Thanks for the suggestion, it was annoying, updated it to use the special casing, seems to be better.